File: src/main/java/app/application/data/entities/YoutubeVideo.java
package app.application.data.entities;

import com.github.kiulian.downloader.model.videos.VideoDetails;
import com.github.kiulian.downloader.model.videos.VideoInfo;
import com.github.kiulian.downloader.model.videos.formats.AudioFormat;
import com.github.kiulian.downloader.model.videos.formats.Format;
import com.github.kiulian.downloader.model.videos.formats.VideoWithAudioFormat;

import java.util.List;

public class YoutubeVideo extends YoutubeEntity<VideoInfo>{

	public YoutubeVideo(VideoInfo videoInfo){
		this.reference = videoInfo;
	}

	public String getVideoId(){
		return getVideoDetails().videoId();
	}

	public String getVideoDescription(){
		return getVideoDetails().description();
	}

	public String getVideoTitle(){
		return getVideoDetails().title();
	}

	public String getVideoThumbnailUrl(){
		return getVideoDetails().thumbnails().get(0).split("\\?sqp")[0];
	}

	public VideoDetails getVideoDetails(){
		return reference.details();
	}

	public List<VideoWithAudioFormat> getVideoWithAudioFormat(){
		return reference.videoWithAudioFormats();
	}

	public AudioFormat getAudioFormat(){
		return reference.audioFormats().get(0);
	}

	public Format getBestVideoWithAudioFormat() {
		return reference.bestVideoWithAudioFormat();
	}
}

File: src/main/kotlin/app/application/data/entities/YoutubeVideo.kt
package app.application.data.entities

import com.github.kiulian.downloader.model.videos.VideoDetails
import com.github.kiulian.downloader.model.videos.VideoInfo
import com.github.kiulian.downloader.model.videos.formats.AudioFormat
import com.github.kiulian.downloader.model.videos.formats.Format
import com.github.kiulian.downloader.model.videos.formats.VideoFormat
import com.github.kiulian.downloader.model.videos.formats.VideoWithAudioFormat
import javafx.scene.image.Image

class YoutubeVideo(videoInfo: VideoInfo?) : YoutubeEntity<VideoInfo?>() {
    init {
        reference = videoInfo
    }

    private var _thumbnailImage: Image? = null
    val thumbnailImage: Image
    get() {
        if(_thumbnailImage == null){
            _thumbnailImage = Image(videoThumbnailUrl)
        }
        return _thumbnailImage as Image
    }
    val videoId: String?
        get() = videoDetails?.videoId()
    val videoDescription: String?
        get() = videoDetails?.description()
    val videoTitle: String?
        get() = videoDetails?.title()
    val videoThumbnailUrl: String?
        get() = videoDetails?.thumbnails()?.get(0)?.split("\\?sqp")?.toTypedArray()?.get(0)
    private val videoDetails: VideoDetails?
        get() = reference?.details()
    val videoWithAudioFormat: MutableList<VideoWithAudioFormat>?
        get() = reference?.videoWithAudioFormats()
    val audioFormat: AudioFormat?
        get() = reference?.audioFormats()?.get(0)
    val bestVideoWithAudioFormat: VideoFormat?
        get() = reference?.bestVideoWithAudioFormat()
}